test/extensions/filters/http/csrf/csrf_filter_integration_test.cc

```diff
@@ -0,0 +1,200 @@
+#include "test/integration/http_protocol_integration.h"
+
+namespace Envoy {
+namespace {
+const std::string CSRF_ENABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+  shadow_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+)EOF";
+
+const std::string CSRF_FILTER_ENABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+)EOF";
+
+const std::string CSRF_SHADOW_ENABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 0
+      denominator: HUNDRED
+  shadow_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+)EOF";
+
+const std::string CSRF_DISABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 0
+      denominator: HUNDRED
+)EOF";
+
+class CsrfFilterIntegrationTest : public HttpProtocolIntegrationTest {
+protected:
+  void testNormalRequest(Http::TestHeaderMapImpl&& request_headers,
+                         const char* expected_response_code) {
+    initialize();
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    auto response = codec_client_->makeRequestWithBody(request_headers, 1024);
+    waitForNextUpstreamRequest();
+    upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, true);
+    response->waitForEndStream();
+
+    ASSERT_TRUE(response->complete());
+    EXPECT_STREQ(expected_response_code, response->headers().Status()->value().c_str());
+  }
+
+  void testInvalidRequest(Http::TestHeaderMapImpl&& request_headers,
+                          const char* expected_response_code) {
+    initialize();
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    auto response = codec_client_->makeRequestWithBody(request_headers, 1024);
+    response->waitForEndStream();
+
+    ASSERT_TRUE(response->complete());
```

Comment:
note that ASSERT_TRUE in a helper method does not actually stop the test in progress. 

In general it's better if you can do the ASSERT/EXPECT macros at the top level rather than in helpers, though EXPECT is OK. The disadvantage is that any failure will report a line-number in the helper which is shared, though if there's only call to the helper method in a single test, it will be reasonably obvious where to look from the logs.

A pattern that would work well here is to eliminate the expected_response_code arg (making my above suggestion moot) and instead returning it, maybe like this:

  return response->complete() ? response->headers().Status()->value().getStringView() : absl::string_view("incomplete");

and then at the call-sites do

  EXPECT_STREQ("403", testInvalidRequest(...));